keps/sig-instrumentation/draft-20190404-kubernetes-control-plane-metrics-stability.md

```diff
@@ -0,0 +1,294 @@
+
+1. the metric itself will not be deleted ([or renamed](#metric-renaming))
+3. the type of metric will not be modified
+4. no labels can be added or removed from this metric
+
+From an ingestion point of view, it is backwards-compatible to add or remove possible __values__ for labels which already do exist (but __not__ labels themselves). Therefore, adding or removing __values__ from an existing label is permissible. Stable metrics can also be marked as __deprecated__ for a future kubernetes version, since this is a metadata field and does not actually change the metric itself.
+
+As an aside, all metrics should be able to be individually disabled by the cluster administrator, regardless of stability class. By default, all non-deprecated metrics will be automatically registered to the metrics endpoint unless explicitly blacklisted via a command line flag (i.e. '--disable-metrics=somebrokenmetric,anothermetric').
+
+## Deprecation Lifecycle
+
+This proposal introduces deprecation metadata for metrics, to be used to define a deprecation lifecycle. Metrics can be annotated with a kubernetes version, from which point that metric will be considered deprecated. This allows us to indicate that a metric is slated for future removal and provides the consumer a reasonable window in which they can make changes to their monitoring infrastructure which depends on this metric. While deprecation policies only actually change stability guarantees for __stable__ metrics (and not __alpha__ ones), deprecation information may however be optionally provided on alpha metrics to help component owners inform users of future intent, to help with transition plans.
+
+When a stable metric undergoes the deprecation process, we are signaling that the metric will eventually be deleted. The lifecyle looks roughly like this (each stage represents a kubernetes release):
+
+__Stable metric__ -> __Deprecated metric__ -> __Hidden metric__ -> __Deletion__
+
+__Deprecated__ metrics have the same stability guarantees of their counterparts. If a stable metric is deprecated, then a deprecated stable metric is guaranteed to *not change*. When deprecating a stable metric, a future kubernetes release is specified as the point from which the metric will be considered deprecated.
+
+```go
+var someCounter = kubemetrics.CounterOpts{
+    Name: "some_counter",
+    Help: "this counts things",
+    StabilityLevel: kubemetrics.STABLE,
+    DeprecatedVersion: "1.15", // this metric is deprecated when the kubernetes version == 1.15
+}
+````
+
+__Deprecated__ metrics will have their description text prefixed with a deprecation notice string '(Deprecated from x.y)' and a warning log will be emitted during metric registration (in the spirit of the official [kubernetes deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli)).
+
+Before deprecation:
+
+```text
+# HELP some_counter this counts things
```

Comment:
I would have expected that somewhere in here is an indicator about the stability of this metric. As the format is typically gzipped, I don't see an issue with having duplicated strings here. Should we add this, I'd expect some form of impact analysis on this regarding network bytes and response latency.

Reply:
We can annotate stable metrics explicitly as well. I'll run perf analysis against the new registry and the old ones and include that analysis in the appropriate PRs.